lib/Parse/ParseDecl.cpp

```diff
@@ -1164,6 +1242,112 @@ Parser::parseDifferentiatingAttribute(SourceLoc atLoc, SourceLoc loc) {
                                   original, linear, params));
 }
 
+/// SWIFT_ENABLE_TENSORFLOW
+/// parseQualifiedDeclName
+///
+///   qualified-decl-name:
+///     type-identifier? unqualified-decl-name
+///   type-identifier:
+///     identifier generic-args? ('.' identifier generic-args?)*
+///
+/// Parses an optional base type, followed by a declaration name.
+/// Returns true on error (if function decl name could not be parsed).
+bool parseQualifiedDeclName(
+    Parser &P, Diag<> nameParseError, TypeRepr *&baseType,
+    DeclNameWithLoc &original) {
+  // If the current token is an identifier or `Self` or `Any`, then attempt to
+  // parse the base type. Otherwise, base type is null.
+  auto currentPosition = P.getParserPosition();
+  bool canParseBaseType = P.canParseTypeIdentifier();
+  P.backtrackToPosition(currentPosition);
+  if (canParseBaseType)
+    baseType = P.parseTypeIdentifier(/*isParsingQualifiedDeclName*/ true)
+    .getPtrOrNull();
+  else
+    baseType = nullptr;
+  
+  // If base type was parsed and has at least one component, then there was a
+  // dot before the current token.
+  bool afterDot = false;
+  if (baseType) {
+    if (auto ident = dyn_cast<IdentTypeRepr>(baseType)) {
+      auto components = ident->getComponentRange();
+      afterDot = std::distance(components.begin(), components.end()) > 0;
+    }
+  }
+  original.Name = P.parseUnqualifiedDeclName(afterDot, original.Loc, nameParseError,
+                                    /*allowOperators*/ true,
+                                    /*allowZeroArgCompoundNames*/ true);
+  // The base type is optional, but the final unqualified decl name is not.
+  // If name could not be parsed, return true for error.
+  if (!original.Name) return true;
+  return false;
+}
+
+ParserResult<TransposingAttr>
+Parser::parseTransposingAttribute(SourceLoc atLoc, SourceLoc loc) {
+  StringRef AttrName = "transposing";
+  SourceLoc lParenLoc = loc, rParenLoc = loc;
+  TypeRepr *baseType;
+  DeclNameWithLoc original;
+  SmallVector<ParsedAutoDiffParameter, 8> params;
+  
+  // Parse trailing comma, if it exists, and check for errors.
+  auto consumeIfTrailingComma = [&]() -> bool {
+    if (!consumeIf(tok::comma)) return false;
+    // Diagnose trailing comma before ')'.
+    if (Tok.is(tok::r_paren)) {
+      diagnose(Tok, diag::unexpected_separator, ",");
+      return true;
+    }
+    // Check that token after comma is 'wrt:'.
+    if (!Tok.is(tok::identifier) || !(Tok.getText() == "wrt")) {
+      diagnose(Tok, diag::attr_transposing_expected_label_linear_or_wrt);
+      return true;
+    }
+    return false;
+  };
+  
+  // Parse '('.
+  if (!consumeIf(tok::l_paren, lParenLoc)) {
+    diagnose(getEndOfPreviousLoc(), diag::attr_expected_lparen, AttrName,
+             /*DeclModifier*/ false);
+    return makeParserError();
+  }
+  
+  {
+    SyntaxParsingContext ContentContext(
+        SyntaxContext, SyntaxKind::DifferentiatingAttributeArguments);
```

Comment:
Please add libSyntax support for `@transposing` attribute. I shared information in [TF-634](https://bugs.swift.org/browse/TF-634) to help you get started.

Reply:
Done here https://github.com/apple/swift/pull/25684/commits/80b70222d9ac02946b0457d8b8f8c73bd20d9c35#diff-4300c9baa89557f9e7d8755b9dce625eR335

Reply:
Done! I made the children the same as `'DifferentiatingAttributeArguments'`. How can I test/verify that it works?

Reply:
Test seems to be failing because of this change it seems:
```
--
found code completion token AVAILABILITY1 at offset 2809
found code completion token AVAILABILITY2 at offset 4131
found code completion token KEYWORD2 at offset 4820
/Users/bartchr/Desktop/Projects/swift-source/swift/test/IDE/complete_decl_attribute.swift:70:18: error: KEYWORD2-NOT: excluded string found in input
// KEYWORD2-NOT: Keyword
                 ^
<stdin>:17:1: note: found here
Keyword/None: transposing[#Func Attribute#]; name=transposing
^~~~~~~

--
```

Reply:
That failure is probably expected.
`test/IDE/complete_decl_attribute.swift` tests code completion for attributes. Since you added an attribute, you have to add a new `// CHECK:` line to the test.

Let me know if that makes sense.